### ISSUE TITLE
Easy-Shared_Folder Feature?

### ISSUE THREAD

**Issue by Seven7zip** [waydroid/waydroid#1044]
**Is your feature request related to a problem? Please describe.**
It's annoying to have to mount it manually every time.
**Describe the solution you'd like**
Can you please add a shared folder feature that will be easy and automatic?

**Comment by TriMoon** [comment]
I agree, they could do that in the startup scripts where the session is started by the user.

**Comment by Seven7zip** [comment]
I tried fstab but it didn't work at all. And this method is too unstable. Some apps may not be able to properly recognize the file.

**Comment by TriMoon** [comment]
@Seven7zip , no idea how/what you tried, but i would suggest you to try mount units using the systemd way...

**Comment by Quackdoc** [comment]
It probably wouldn't be hard to get waydroid to work with post-up scripts for this. however it's not something I can work on, systemd is also a possible solution

**Comment by FalconSN** [comment]
Is there an example on how to make a systemd configuration? I have tried fstab entry with x-systemd.automount option, timing it out and with systemd way with just "WantedBy=multi-user.target" condition. fstab entry actually did work, but the mountpoint stayed only in /data/media/0/Download and didn't reflect to storage/emulated/0, which was the closest I could get without mount --bind.

If not, please help me to realize the mistakes in this small script, because I couldn't make it work:
```
#! /bin/bash
readarray -t lines <(waydroid "$@")
        for line in "${lines[@]}"; do
                if [[ "$line" == *"ready"* ]]; then
                        echo "Mounting"
                        sudo mount --bind "/home/falconsn/Downloads" "/home/falconsn/.local/share/waydroid/data/media/0/Download"
                fi
        done
```

**Comment by TriMoon** [comment]
@FalconSN, To easily generate a systemd-mount unit try this:
- :information_source: _This code is **not tested**, it was written online as a reply..._
   ```bash
   #!/usr/bin/env bash
   #       SPDX-License-Identifier: CC-BY-NC-SA-4.0
   #
   srcdir="~/Downloads"
   dstdir="~/.local/share/waydroid/data/media/0/Download"
   sudo systemd-mount -o bind "${srcdir}" "${dstdir}"

   unitfile=$(systemd-escape --suffix=mount --path "${dstdir}")
   systemctl cat "${unitfile}"
   ```
That generated unit-file can be copied over to `/etc/systemd/system/` so you can use it again later when needed.
F.e. in your script you could replace the respective portion with:
```bash
echo "Mounting"
dstdir="~/.local/share/waydroid/data/media/0/Download"
unitfile=$(systemd-escape --suffix=mount --path "${dstdir}")
sudo systemctl start "${unitfile}"
```

_(Unfortunatly you can't use mount units in the systemd user manager, because it needs root.
Hence the usage of `sudo` or `pkexec` is required...)_

**Comment by FalconSN** [comment]
@TriMoon, I'm sorry for late reply.
I have created a small script that starts waydroid with while loop which mounts the particular folders with the condition of "ready" string and then breaks. Unfortunately, this configuration is as you said, simple and does not provide any solution for my problem which is; (not) mounting the folders on Waydroid boot and unmounting on Waydroid shutdown. Apologizes for not being clear in the previous comment. I really would love to work on making Python script for this, but unfortunately don't have the knowledge to do so.

**Comment by TriMoon** [comment]
> does not provide any solution for my problem which is; (not) mounting the folders on Waydroid boot and unmounting on Waydroid shutdown.

1. You could create "Drop-in" a fragment(s) for the `waydroid-container.service`, using either `Requires=` or `Wants=` in combination with `After=` for your service that bind-mounts those directories.
2. Or use `RequiredBy=` or `WantedBy=` in the `[Install]` section, in combination with a `Before=` in the `[Unit]` section, of your service that bind-mounts those directories.

For example i'm using this drop-in to clear the log files of waydroid before it starts:
```ini
# /etc/systemd/system/waydroid-container.service.d/clear-logs.conf
[Service]
ExecStartPre=rm -f /var/lib/waydroid/waydroid.log
```

**Comment by FalconSN** [comment]
> 1. You could create (a) "Drop-in" fragment(s) for the `waydroid-container.service`, using either `Requires=` or `Wants=` in combination with `After=`, in the `[Unit]` section, for your service that bind-mounts those directories.

Using `waydroid-container.service` won't work. What I meant by 

> (not) mounting the folders on Waydroid boot and unmounting on Waydroid shutdown.

is session starts and stops. For some reason, mounting the directories before "Android with user 0 is ready" output (in other words; before booting is complete) is not working well; mounts stay in /data/media/0 and not being reflected to /storage/emulated/0. For this particular problem, I had made a `waydroid-session.service` with `Upholds=` option that points to a mount unit, but this didn't work either because there is no way to delay it. It might be possible to, though, by starting the mount unit with `ExecStartPost=bash 'sleep 30 ; sudo systemctl start <mountunit>'` in `waydroid-session.service`. However, I have discarded this possible solution because it will slowdown the system by starting a waydroid session on boot, and the problem of "not unmounting on service stop" will not go away.

Also, the reason why automount is not working is because; while waydroid is making some complicated mounts in /var/lib/waydroid, it's accessing to `.local/share/waydroid/data/media...` directories and automount is being mounted; before boot up process is done.

I will probably stick with my temporary solution until there's a built-in support or a guide for making configurations for this.

**Comment by stiller-leser** [comment]
Anyone made progress on this?

I've tried a systemd service like this:

```
[Unit]
Description=Waydroid Mount Service
After=waydroid-container.service
Requires=waydroid-container.service

[Service]
Type=oneshot
ExecStart=/usr/bin/mount --bind ... ...
RemainAfterExit=true

[Install]
WantedBy=multi-user.target
```

But the folders remain empty...

**Comment by FalconSN** [comment]
The solution I came up with back in time is, writing a script and starting waydroid session through it.

```sh
#!/usr/bin/bash
[[ "$1 $2" == "session start" || "$1" == "show-full-ui" ]] || exec /usr/bin/waydroid "$@";
id=$(id -u);
[ $id = 0 ] || exec sudo --preserve-env env _USER_=$USER "$(readlink -f "$0")" "$@"
waydroid="$HOME/.local/share/waydroid/data/media/0"
dirs_to_mount=("$HOME/Downloads=$waydroid/Download" # = is delimiter
               "$HOME/Pictures=$waydroid/Pictures")

unmount_on_exit() {
    echo "Umounting partitions..." ;
    for d in "${dirs_to_mount[@]}"; do
        d="${d#*=}";
        umount "$d" && echo "Success: $d" || echo "Fail: $d" >&2;
    done
}

trap unmount_on_exit EXIT

output=".*Android with user [0-9]+ is ready.*" ;

while read -r line; do
    echo "$line"
    if [[ "$line" =~ $output ]]; then
        for d in "${dirs_to_mount[@]}"; do
            src="${d%%=*}";
            dest="${d#*=}";
            if ! [ -d "$dest" ]; then
                parent="${dest%/*}";
                if [ -d "$parent" ]; then
                    owner="$(stat -c '%u' "${dest%/*}")";
                    if mkdir -p "$dest"; then
                        chown $owner:$owner "$dest"
                    else
                        echo "Couldn't create: $dest" >&2;
                        continue;
                    fi
                else
                    echo "destination directory not found: $dest" >&2;
                    continue;
                fi
            fi
            echo -n "Mounting $dest... ";
            bindfs --perms=755 --create-for-user=$_USER_ --create-with-perms=755 --map=$_USER_/10136:@$_USER_/@10136 "$src" "$dest" && \
                echo "success." || echo "fail." >&2;
        done
    fi
done < <(sudo --preserve-env --user=$_USER_ /usr/bin/waydroid "$@" 2>&1)
```

**Comment by ayasa520** [comment]
I added the shared folder feature to [waydroid-helper](https://github.com/ayasa520/waydroid-helper), which should be easy to use.

It uses a user-level systemd service and a system-level D-Bus service to monitor Waydroid logs via inotify instead of polling. When "Android with user 0 is ready" is detected, it triggers the mounting, ensuring the shared folder works. Initially, I intended to achieve this with gbinder, but it interfered with Waydroid's service, so I decided to stick with log monitoring.
